libraries/AP_Compass/AP_Compass_MMC5xx3.cpp

```diff
@@ -0,0 +1,297 @@
+     */
+    switch (state) {
+
+    // perform a set operation
+    case STATE_SET: {
+        if (!dev->write_register(REG_CONTROL0, REG_CONTROL0_SET)) {
+            break;
+        }
+        // minimum time to wait after set/reset before take measurement request is 1ms
+        state = STATE_SET_MEASURE;
+        break;
+    }
+
+    // request a measurement for field and offset calculation after set operation
+    case STATE_SET_MEASURE: {
+        if (!dev->write_register(REG_CONTROL0, REG_CONTROL0_TM)) {
+            break;
+        }
+        state = STATE_SET_WAIT;
+        break;
+    }
+
+    // wait for measurement to be ready after set operation, then read the
+    // measurement data and request a reset operation
+    case STATE_SET_WAIT: {
+        uint8_t status;
+        if (!dev->read_registers(REG_STATUS, &status, 1)) {
+            state = STATE_SET;
+            break;
+        }
+
+        // check if measurement is ready
+        if (!(status & 1)) {
+            break;
+        }
+
+        // read measurement
+        if (!dev->read_registers(REG_XOUT_L, (uint8_t *)&data0[0], 6)) {
+            state = STATE_SET;
+            break;
+        }
+
+        // request set operation
+        if (!dev->write_register(REG_CONTROL0, REG_CONTROL0_RESET)) {
+            break;
+        }
+        // minimum time to wait after set/reset before take measurement request is 1ms
+        state = STATE_RESET_MEASURE;
+        break;
+    }
+
+    // request a measurement for field and offset calculation after reset operation
+    case STATE_RESET_MEASURE: {
+        // take measurement request
+        if (!dev->write_register(REG_CONTROL0, REG_CONTROL0_TM)) {
+            state = STATE_SET;
+            break;
+        }
+
+        state = STATE_RESET_WAIT;
+        break;
+    }
+
+    // wait for measurement to be ready after reset operation,
+    // then read the measurement data, calculate the field and offset,
+    // and begin requesting field measurements
+    case STATE_RESET_WAIT: {
+        uint8_t status;
+        if (!dev->read_registers(REG_STATUS, &status, 1)) {
+            state = STATE_SET;
+            break;
+        }
+
+        // check if measurement is ready
+        if (!(status & 1)) {
+            break;
+        }
+
+        uint16_t data1[3];
+        if (!dev->read_registers(REG_XOUT_L, (uint8_t *)&data1[0], 6)) {
+            state = STATE_SET;
+            break;
+        }
+
+        Vector3f field;
+
+        /*
+          calculate field and offset
+         */
+        Vector3f f1(float(data0[0]) - zero_offset,
+                    float(data0[1]) - zero_offset,
+                    float(data0[2]) - zero_offset);
+        Vector3f f2(float(data1[0]) - zero_offset,
+                    float(data1[1]) - zero_offset,
+                    float(data1[2]) - zero_offset);
+
+        field = (f1 - f2) * (counts_to_milliGauss / 2);
+        Vector3f new_offset = (f1 + f2) * (counts_to_milliGauss / 2);
```

Comment:
Should be *0.5

Reply:
to be clear:
```
        field = (f1 - f2) * counts_to_milliGauss * 0.5f;
        Vector3f new_offset = (f1 + f2) * counts_to_milliGauss  * 0.5f;
```

this is the suggestion?

Reply:
because multiplication is faster than division

Reply:
done